Rowing Machine Stroke Rate Spikes

DXCanas opened the pull request:
Noticing some insane spikes in my workouts, thought others might appreciate something like this. But I skipped the Issue flow — hope that's alright.

DXCanas committed:
Adding sanity check to cadence write for rowing.

Using > max olympic stroke rate.

MrCsabaToth commented:
We must not hard wire constant literals like 50: it's an anti pattern called "magic numbers". (I don't say that we don't have any magic numbers right now, but I try to avoid them). At minimum I try to collect them into thematic constants.dart like files, but it should be perhaps configurable. Because I checked and in certain cases over 50 spm can be plausible.

https://share.google/aimode/xjUk7dQViAL0zfTuR

Brian Shaw rowing sprint https://share.google/KWWzgTuKhbvfiD288

MrCsabaToth commented:
I'm on an airplane and it's hard to check, but some Bluetooth profiles like RSC and CSC don't provide computed cadence, but rather just turnaround events and their time stamps. In such a case I need to assemble an event queue and compute cadence over the sliding time window. During such calculations overflows occur (unavoidably), and in general I am prepared for those, but such overflow could cause spikes if a calculation is too naive. (Extreme spikes, like truly close to max).
I cannot remember if the Rower Data profile within FTMS provides computed cadence or not. In case we are not computing it, then it means that the rower console's micro controller performs the similar computations I outlined and it can make overflow mistakes. Do you see spikes on the rower console? Maybe the firmware only makes glitches when it converts the data to the FTMS Rower Data format for transmission?

DXCanas commented:
> it's an anti pattern called "magic numbers".

Fair! I went in to see why the smoothing window wasn't being applied and found the history behind the `if` in the blame. Too quick and dirty 🙃 

> Because I checked and in certain cases over 50 spm can be plausible.

Did a quick search for "highest recorded rowing cadence olymics" and rounded up to the nearest 10th. Not very scientific 😅 

> I cannot remember if the Rower Data profile within FTMS provides computed cadence or not.

[It gives stroke rate and, iirc, you use the cadence column to hold it.](https://bitbucket.org/bluetooth-SIG/public/raw/8fd0701c91777053b83cf8aa27e4c65761e65dfb/gss/org.bluetooth.characteristic.rower_data.yaml)

> Do you see spikes on the rower console? Maybe the firmware only makes glitches when it converts the data to the FTMS Rower Data format for transmission?

Honestly, I haven't put in a long enough session with just the rower console to validate — the monitor on this machine is notorious, and having the data recorded was more important to me. 

This is what the spikes look like:
<img width="725" height="379" alt="image" src="https://github.com/user-attachments/assets/97053304-ac72-49e2-80d1-0b97c2982c7c" />

They tend to align roughly with the "sprint" portion of intervals. Would some debug logs help diagnose, if we're moving away from magic numbers?

MrCsabaToth commented:
That graph is extremely helpful. The spikes are jarring, but that's a good sign because we could use a much higher threshold than 50, probably even 100 to filter those out.
However they also gave me a clue what might be happening: https://share.google/aimode/wexAVC0JzgU0qEkpF

> The Bluetooth Fitness Machine Service (FTMS) standard uses specific values to indicate that a data field is **not available** or **not set**.
> These special values vary depending on the data type of the field within the characteristic:
> * For **integer** data types (e.g., `SINT16`, `UINT16`, `SINT24`, `UINT24`, `UINT8`), a value of `0xFF` (or the equivalent for multi-byte integers, e.g., `0xFFFF` for `UINT16`) is typically used to indicate "not available".
> * For **float** data types (e.g., `FLOAT16`, `FLOAT32`), the IEEE 754 standard special value for **Not a Number (`NaN`)** is used to signal "not available". 

In other words I think we get 255, and since the resolution of the stroke rate is 0.5, this is interpreted as 127.5 instead of N/A / NaN.

A debug log is needed to verify that and then examination of the core logic, because this (we don't respect N/A magic values) could be an issue elsewhere too.

DXCanas committed:
Reverting commit due to discussion in PR.
Seems to be indicative of a broader reaching problem that is worth addressing properly.
No magic numbers.

This reverts commit 2e5de91e.

DXCanas commented:
Sifting through logs a bit, because I left `INFO` level on too long 😅 

But with some basic filtering, I am seeing behaviour like you described:
```
2025-12-30 13:25:48.337872	INFO	FITNESS_EQUIPMENT	mergedToYield	merged id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.337789 | distance 2488.0 | elapsed 827 | calories 137 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.338585	INFO	FITNESS_EQUIPMENT	processRecord	stub at 2025-12-30 13:25:48.338536 id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.338519 | distance 2488.0 | elapsed 827 | calories 137 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.339801	INFO	FITNESS_EQUIPMENT	processRecord	adjusted stub id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.338519 | distance 2488.0 | elapsed 827 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.341084	INFO	FITNESS_EQUIPMENT	processRecord	stub before cumulative id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:48.338519 | distance 2488.0 | elapsed 766 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 766127 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.341669	INFO	FITNESS_EQUIPMENT	processRecord	stub after processable id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:48.338519 | distance 2488.0 | elapsed 766 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 766127 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.821740	INFO	FITNESS_EQUIPMENT	mergedToYield	merged id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.821661 | distance 2488.0 | elapsed 827 | calories 137 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.822564	INFO	FITNESS_EQUIPMENT	processRecord	stub at 2025-12-30 13:25:48.822495 id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.822474 | distance 2488.0 | elapsed 827 | calories 137 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.824431	INFO	FITNESS_EQUIPMENT	processRecord	adjusted stub id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:48.822474 | distance 2488.0 | elapsed 827 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.827864	INFO	FITNESS_EQUIPMENT	processRecord	stub before cumulative id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:48.822474 | distance 2488.0 | elapsed 767 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 766611 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:48.828562	INFO	FITNESS_EQUIPMENT	processRecord	stub after processable id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:48.822474 | distance 2488.0 | elapsed 767 | calories 137 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 766611 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.307401	INFO	FITNESS_EQUIPMENT	mergedToYield	merged id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.307173 | distance 2492.0 | elapsed 828 | calories 138 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.308317	INFO	FITNESS_EQUIPMENT	processRecord	stub at 2025-12-30 13:25:49.308252 id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.308229 | distance 2492.0 | elapsed 828 | calories 138 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.309880	INFO	FITNESS_EQUIPMENT	processRecord	adjusted stub id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.308229 | distance 2492.0 | elapsed 828 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.311554	INFO	FITNESS_EQUIPMENT	processRecord	stub before cumulative id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:49.308229 | distance 2492.0 | elapsed 767 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 767096 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.312329	INFO	FITNESS_EQUIPMENT	processRecord	stub after processable id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:49.308229 | distance 2492.0 | elapsed 767 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 767096 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.792291	INFO	FITNESS_EQUIPMENT	mergedToYield	merged id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.792208 | distance 2492.0 | elapsed 828 | calories 138 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.793078	INFO	FITNESS_EQUIPMENT	processRecord	stub at 2025-12-30 13:25:49.793043 id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.793025 | distance 2492.0 | elapsed 828 | calories 138 | power 197 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.794328	INFO	FITNESS_EQUIPMENT	processRecord	adjusted stub id -9223372036854775808 | activityId -9223372036854775807 | timeStamp 2025-12-30 13:25:49.793025 | distance 2492.0 | elapsed 828 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis null | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.795546	INFO	FITNESS_EQUIPMENT	processRecord	stub before cumulative id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:49.793025 | distance 2492.0 | elapsed 768 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 767581 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0			
2025-12-30 13:25:49.796132	INFO	FITNESS_EQUIPMENT	processRecord	stub after processable id -9223372036854775808 | activityId 16 | timeStamp 2025-12-30 13:25:49.793025 | distance 2492.0 | elapsed 768 | calories 138 | power 270 | speed 14.87603305785124 | cadence 27 | heartRate 173 | elapsedMillis 767581 | pace 121.0 | caloriesPerHour 0.0 | caloriesPerMinute 0.0 | resistance 4 | preciseCadence 27.0 | strokeCount 255.0
```			
I'll try and get you a file in a moment

DXCanas commented:
Alright, I've cut off anything without that sprint workout's date! The larger one had several days' worth and it was… unruly.
[Filtered-2025-12-30-DebugLogs.csv.zip](https://github.com/user-attachments/files/24402638/Filtered-2025-12-30-DebugLogs_20260101T185630844394Z.csv.zip)

DXCanas committed:
Handle FTMS "not available" sentinel values regardless of optional flag.
According to FTMS spec, sentinel values (`0xFF` for `UINT8`, `0xFFFF` for `UINT16`, `0xFFFFFF` for `UINT24`) indicate "not available".
They should always return null, regardless of whether the field is marked as optional.
Previously these checks were only performed when optional=true.

DXCanas committed:
Simplifying test for null data from machine.

DXCanas commented:
Took a stab at it! No hard feelings if wrong. This has all been very educational for me.

MrCsabaToth commented:
I need to track down the `optional` boolean: if we'd declare this row cadence byte as optional, then the N/A logic would apply. The fix essentially applies the N/A value check regardless of optional or not. This means `null` can be a result any time (even when non optional), and we need to think if that can have any consequences in the code (where it doesn't expect that).
Must note that a good portion of this code was created when Dart nullability didn't exist (or right when it was introduced as a language feature), and this nullability should at least guarantee that there won't be any crash or null dereference. But I still want to think a little, like what does optional even mean if we don't use it, we can eliminate it.

DXCanas commented:
Yeah! I saw it being used in a few other places — not exclusively for data validity checking — after I essentially performed a fix adding optional to most fields. 

I assumed it was doing more after seeing the places it was checked and didn’t want to dig deeper; didn’t want to break assumptions, or water down the optional nature of fields across the code base 🤷

DXCanas commented:
So, something I ran into during a workout yesterday that might help in the `optional` bool sleuthing:

I tried using TMIW on an elliptical in my gym — I had tracked a small session before, and it had all the important values. So I figured why not. 

This time around, the values were all mucked up:
<img width="500" alt="pre-reboot metrics malfunction" src="https://github.com/user-attachments/assets/67b1d347-d6ba-4caa-8023-cdce5998078f" />
On the left is the tracking of the session while I was on the machine. On the right is an unpacked fit file from the last (test) sesssion (where I was not wearing a GymLink compatible HR monitor or using the contact strips)

Later, when parsing, I saw that the fit/tcx file has `0` entries for distance, speed, and vertical speed. Only values were HR and Power. Cadence was there, but set to 254 for the entire workout. 

This morning, I went back. Replicated. As the gym was empty enough, I rebooted the elliptical. 

After a few attempts, I was able to reconnect, and I got all the values, nice and clean:
<img width="500" alt="post-reboot metrics functioning" src="https://github.com/user-attachments/assets/34cf9f10-6110-4bdb-8d69-ce1a20058359" />

If deepwiki isn't lying to me, the non-optional metrics for elliptical/cross-trainer are:
- Instant Speed
- Total Distance
- Instant Power
- Elapsed Time

It seems that, pre-reboot, the elliptical was bugging out and not reporting at _least_ distance or speed. TMIW seemed to be aware of this:
<img width="500" alt="pre-reboot features" src="https://github.com/user-attachments/assets/6cc4d711-2252-4cce-a65c-55adc8dcb659" />

And yet the display was munged, with the data only a little better.

All this to say: Maybe the `optional` flag influences display in some way? It already seems to have some effect on data writing, as we've established.

MrCsabaToth commented:
That's weird. The `optional` should not modify the byte layout, seemingly it only controls if the particular measurement respects the sentinel values to provide N/A (null) data, or not. This flag was introduced almost 5 years ago, so my memory faded, but the answer below reinforce my own perception. Here is a part of the chat with my agent (antigravity + Gemini 3.0 Pro):

"Here are the answers to your questions regarding the `optional` flag in `MetricDescriptor`:

1. **When was the optional flag introduced?** It was introduced on **March 18, 2021** (commit date for the creation of `lib/devices/metric_descriptors/metric_descriptor.dart`).
2. **Why not all the sentinel values always handled to abide by FTMS (or broader) standard?** The implementation is "opt-in" to provide flexibility. While the FTMS standard defines maximum values (e.g., `0xFF`, `0xFFFF`) as "Data Not Available," not all devices or metrics in the codebase might strictly follow this, or they might reuse those values for valid data in proprietary contexts. The `optional: true` flag allows the developer to explicitly enforce this FTMS behavior only where appropriate, avoiding potential data loss for non-standard devices.
3. **Can null returned values cause any unexpected effect? No.** The `DataHandler` class (which consumes these descriptors) explicitly returns nullable types (e.g., `double? getSpeed(...)`). This means the nullability is propagated up to the caller (UI or logic), which is forced by Dart's type system to handle the `null` case (usually by displaying `"--"` or N/A).
4. **Since Dart language gained nullability feature, does optional flag make sense at all? Yes, it is still necessary.** The optional flag and Dart's null safety serve different purposes:
   * **Optional Flag (Runtime)**: Handles the decoding logic—it interprets a specific binary value (e.g., `0xFFFF`) as "no data". Without this flag, the system would interpret `0xFFFF` as the number `65535`.
   * **Dart Null Safety (Compile-time)**: Handles the *resulting state*—it ensures that once the value is decoded as `null`, the code forces you to handle that absence safely. You need the flag to tell the system when to produce the `null` that Dart captures.
5. **Why do you think it is named "optional"?** The name likely refers to the value's presence, not the field's presence. Even if a field is included in the packet (based on the flag bits), the value itself can still be "optional" (indicated by the sentinel). It essentially means "Value Can Be Absent".

DXCanas commented:
Hm alright — I'll keep an eye out for a decision then? Let me know if there's anything I can do to help debug. I did also forget to attach debug logs:
[DebugLogs_20260105T173232436763Z.zip](https://github.com/user-attachments/files/24441508/DebugLogs_20260105T173232436763Z.zip)

MrCsabaToth commented:
Does that debug log contain both the faulty behavior you described before the reboot and then after the reboot?

I noticed with the LifeFitness ellipticals (and also recumbent bikes, upright bikes, treadmills, and stair climbers), that they report different flag combinations in warmup and cool down mode vs during the actual workout. Maybe our flag interpretation has a bug (happened before, I'll look for a ref), or the machine doesn't abide by FTMS (like Mr Captain rowers, we need to look up if we have a workaround for True Fitness and maybe they fixed a bug in the firmware we don't have to workaround any more?).

BTW Life Fitness is one which reports Heart rate even when not paired, but in that case it sends the sentinel values.

I'm leaning towards eliminating the optional flag, but that will be in a separate PR and issue (apart from the one needed for your rower spm). Theoretically we just need to set optional = true for that bit flag to fix the spm. I'll think about this more.

DXCanas commented:
> Does that debug log contain both the faulty behavior you described before the reboot and then after the reboot?

Should do, yeah! There are 3 sessions in there, I think. The last one was the neat one

DXCanas committed:
Make FTMS sentinel value checking consistent across all MetricDescriptor types.

LongMetricDescriptor and SixByteMetricDescriptor now check sentinel values unconditionally.
Matches the behavior of ByteMetricDescriptor, ShortMetricDescriptor, and ThreeByteMetricDescriptor.

DXCanas committed:
Remove optional parameter from MetricDescriptor classes

FTMS sentinel values are now always checked unconditionally per spec.
The optional parameter is no longer needed since sentinel value detection
does not depend on whether a field's presence is flag-dependent.

- Remove optional from MetricDescriptor base class
- Remove optional from all subclasses (Byte, Short, ThreeByte, Long, SixByte)
- Remove optional: true usages in fitness_machine_descriptor and mr_captain_descriptor
- Update all tests to reflect the simplified API

MrCsabaToth commented on lib/devices/metric_descriptors/long_metric_descriptor.dart:
```diff
@@ -15,7 +16,11 @@ class LongMetricDescriptor extends MetricDescriptor {
-    if (optional && value == maxUint32 - 1) {
+    // FTMS spec: 0xFFFFFFFF (4294967295) indicates "not available" for UINT32 fields
+    if (value == maxUint32 - 1) {
+      debugPrint(
```
Not 100% sure if we need a debugPrint. For debugging certainly OK, and I see that the PR is draft.

DXCanas replied:
Should we get rid of it before merge? Or after?

MrCsabaToth replied:
I can take care of it after.

MrCsabaToth commented on test/six_byte_metric_descriptor_test.dart:
```diff
@@ -23,52 +23,49 @@ void main() {
     final rnd = Random();
-    for (var lenMinusFive in getRandomInts(repetition, 99, rnd)) {
-      final len = lenMinusFive + 5;
+    for (var lenMinusSix in getRandomInts(repetition, 99, rnd)) {
```
Interesting catch. Maybe several copy-paste errors (originally( here from the three byte test and the AI found that?

MrCsabaToth commented on test/three_byte_metric_descriptor_test.dart:
```diff
@@ -61,10 +57,9 @@ void main() {
+          expect(desc.getMeasurementValue(data), expected == null ? null : closeTo(expected, eps));
         },
       );
     }
```
I'd also want to have a two tests from the recordings you have. Picking a few data packets when there wasn't a spike, and a few packets where there was a spike and check those against our verified perceived / expected outcome records.
Example test with actual measurements, one example of many https://github.com/TrackMyIndoorWorkout/TrackMyIndoorWorkout/blob/develop/test/life_fitness_treadmill_test.dart
I'd name the test file by the name of brand / model, since those packets are coming from there.

DXCanas replied:
being honest, now is not a great time for me to implement this; followup PR, for some time in the future? 

Also: I don't think I have the sorts of recording files that would be fed in through that mechanism? Unless TMIW created them in the background.

MrCsabaToth replied:
Oh the tests I mentioned would not be for the six bytes FTMS data, I haven't seen such in any FTMS standards I can remember. The tests I meant for the rower and elliptical, but I can also do that later.

MrCsabaToth commented on intention.md:
```diff
@@ -0,0 +1,492 @@
+So, this morning, I wrote this issue about this project this is stored in:
```
I know this is a Draft PR. In other projects where I use AI extensively I like to establish a "memories" directory. Unless the documentation is generic, for the user / developer it can go into the docs. Otherwise it could be beneficial to keep around files about AI planning, decisions, implementation, etc. In the "memories" I organize files in a way that it starts with the date YYYY-MM-DD_{feature_name, other addition like plan, etc}.md. This way it'll be ordered by date and many years from now we can prune them. It's also good if the AI leaves meta-data in the file as well regarding date, time, etc.

DXCanas replied:
Oh, ha 😅 this wasn’t supposed to be included. But. Will read and reply your comment more in depth later

MrCsabaToth commented:
There's a code formatting change which keeps this PR from a mergeable state. This PR is valuable, it'd be great to merge

DXCanas commented:
rebased! and added the formatting changes in a separate commit :)

Edit: Ah, still failing. Will update local flutter.

DXCanas committed:
chore: reformat code with dart format

Apply automatic code formatting across MetricDescriptor classes and related files.
No functional changes.

MrCsabaToth commented:
Have you noticed any buggy behavior / regression with the elliptical or other fitness machines? I'll test it too, but you seem to have access to more machines as well

MrCsabaToth reviewed (approved):
LGTM with later tasks - not to myself -: remove `debugLog` (or convert to proper logging) and adding tests for devices in the logged data (two `DebgugLog`s attached to the linked issue)

DXCanas commented:
> Have you noticed any buggy behavior / regression with the elliptical or other fitness machines? I'll test it too, but you seem to have access to more machines as well

Being honest, I’ve been using other FTMS apps or recording methods. But I do have access to FTMS compatible:
- Treadmill 
- Elliptical
- Stairmaster
- Rower (though I haven’t been using that monitor)

can run some tests if we can pin down procedures!

MrCsabaToth commented:
* Oh, you don't have to be go out of your way, I was just remembering that at one point along the changes you reported some glitches in one of the fitness machines. And I was wondering if that's resolved.
* Test would be basically to verify along a short workout (or workout length of your choice) that the metrics displayed by the app match or are very close to the values displayed on the fitness machine consoles, or in case of lack of console they are close to the expected values.
* I haven't tested your changes yet, but I certainly will, I need to rebase my own fork.
